src/webgl/shader.js

```diff
@@ -16,5 +19,286 @@ module.exports = {
   lightVert:
     fs.readFileSync(__dirname + '/shaders/light.vert', 'utf-8'),
   lightTextureFrag:
-    fs.readFileSync(__dirname + '/shaders/light_texture.frag', 'utf-8')
-};
\ No newline at end of file
+    fs.readFileSync(__dirname + '/shaders/light_texture.frag', 'utf-8'),
+};
+
+p5.Shader = function(renderer, vertId, fragId) {
+  // TODO: adapt this to not take ids, but rather,
+  // to take the source for a vertex and fragment shader
+  // to enable custom shaders at some later date
+  this._renderer = renderer;
+  this._vertId = vertId;
+  this._fragId = fragId;
+  this._vertShader = -1;
+  this._fragShader = -1;
+  this.glProgram = 0;
+  this.loadedAttributes = false;
+  this.attributes = {};
+  this.loadedUniforms = false;
+  this.uniforms = {};
+  this.bound = false;
+
+  return this;
+};
+
+p5.Shader.prototype.init = function() {
+  if (this.glProgram === 0 /* or context is stale? */) {
+    var gl = this._renderer.GL;
+
+    //set up our default shaders by:
+    // 1. create the shader,
+    // 2. load the shader source,
+    // 3. compile the shader
+    this._vertShader = gl.createShader(gl.VERTEX_SHADER);
+    //load in our default vertex shader
+    gl.shaderSource(this._vertShader, shaders[this._vertId]);
+    gl.compileShader(this._vertShader);
+    // if our vertex shader failed compilation?
+    if (!gl.getShaderParameter(this._vertShader, gl.COMPILE_STATUS)) {
+      alert('Yikes! An error occurred compiling the vertex shader:' +
+        gl.getShaderInfoLog(this._vertShader));
+      return null;
+    }
+
+    this._fragShader = gl.createShader(gl.FRAGMENT_SHADER);
+    //load in our material frag shader
+    gl.shaderSource(this._fragShader, shaders[this._fragId]);
+    gl.compileShader(this._fragShader);
+    // if our frag shader failed compilation?
+    if (!gl.getShaderParameter(this._fragShader, gl.COMPILE_STATUS)) {
+      alert('Darn! An error occurred compiling the fragment shader:' +
+        gl.getShaderInfoLog(this._fragShader));
+      return null;
+    }
+
+    this.glProgram = gl.createProgram();
+    gl.attachShader(this.glProgram, this._vertShader);
+    gl.attachShader(this.glProgram, this._fragShader);
+    gl.linkProgram(this.glProgram);
+    if (!gl.getProgramParameter(this.glProgram, gl.LINK_STATUS)) {
+      alert('Snap! Error linking shader program');
+    }
+
+    this.loadAttributes();
+    this.loadUniforms();
+
+    // TODO move elsewhere?
+    this._renderer._createEmptyTexture();
```

Comment:
maybe for now a comment mentioning this only happens once, so a developer doesn't have to look up the _createEmptyTexture() definition to confirm that.